h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,139 @@
+package ai.h2o.automl.targetencoding;
+
+import water.MRTask;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.util.IcedHashMap;
+
+class BroadcastJoinForTargetEncoder {
+  
+  static class CompositeLookupKey {
+    // Note: Consider using indexes of levels instead of strings to save space. It will probably require 
+    // to use CategoricalWrappedVec.computeMap() since indexes for levels could differ in both frames for the same level values.
+    private String _levelValue;
+    private Long _foldValue;
+
+    CompositeLookupKey(String levelValue, long fold) {
+      this._levelValue = levelValue;
+      this._foldValue = fold;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj != null && obj instanceof CompositeLookupKey) {
+        CompositeLookupKey s = (CompositeLookupKey) obj;
+        return _levelValue.equals(s._levelValue) && _foldValue.equals(s._foldValue);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return (_levelValue + _foldValue).hashCode();
```

Comment:
Creating a new object (String) to calculate a hashCode is very inefficient. I strongly suggest to change it. A simple auto-generated hashCode should be good enough for your purposes and provide good performance.

Reply:
@deil87 I didn't mean the default java hash code, that won't work. I was refering to an auto-generated hash code that IntelliJ will give you (or any other IDE). Eg.:

    @Override
    public int hashCode() {
      int result = _levelValue.hashCode();
      result = 31 * result + _foldValue.hashCode();
      return result;
    }